x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java

```diff
@@ -142,18 +164,21 @@ protected final boolean assertCurrentThreadMayLoadSnapshot() {
      * @return true if the snapshot was loaded by executing this method, false otherwise
      */
     public boolean loadSnapshot() {
+        assert assertCurrentThreadMayLoadSnapshot();
         boolean alreadyLoaded = this.loaded;
         if (alreadyLoaded == false) {
             synchronized (this) {
                 alreadyLoaded = this.loaded;
                 if (alreadyLoaded == false) {
                     this.blobContainer = blobContainerSupplier.get();
                     this.snapshot = snapshotSupplier.get();
+                    if (loadCacheEagerly) {
+                        prewarmCache();
```

Comment:
This methods blocks until the cache is fully prewarmed. It must be done before `loaded` is set to `true` so that other components of the system are not likely to trigger some caching on this directory files.

Reply:
Is that strictly necessary? I would prefer to initiate the prewarming here, but at the same time allow the shard routing to move to started state as quickly as possible.

Reply:
This is not strictly necessary but a misunderstanding from my part. We discussed this and I updated the PR so that cache warming now runs concurrently with the recovery.